File: product/src/main/java/com/mackittipat/microshop/product/service/CategoryService.java
package com.mackittipat.microshop.product.service;

import com.mackittipat.microshop.product.dto.CategorySearchForm;
import com.mackittipat.microshop.product.dto.CategorySearchResult;
import com.mackittipat.microshop.product.entity.Category;

import java.util.List;

public interface CategoryService {

  List<Category> findAll();

  CategorySearchResult search(CategorySearchForm categorySearchForm);

  Category findById(long id);

  void create(Category category);

  void update(Category category);
}

File: web-admin/src/app/services/category.service.ts
import {Injectable} from '@angular/core';
import {CategorySearchForm} from "../dto/category-search-form";
import {HttpClient, HttpParams} from "@angular/common/http";
import {environment} from '../../environments/environment';
import {Observable} from "rxjs";
import {CategorySearchResult} from "../dto/category-search-result";
import {Category} from "../dto/category";

@Injectable({
  providedIn: 'root'
})
export class CategoryService {

  constructor(private httpClient: HttpClient) {
  }

  search(categorySearchForm: CategorySearchForm): Observable<CategorySearchResult> {
    return this.httpClient.get<CategorySearchResult>(
      environment.url_product + 'categories', {
        params: new HttpParams()
          .set('page', categorySearchForm.page)
          .set('name', categorySearchForm.name)
      });
  }

  findById(id: number): Observable<Category> {
    return this.httpClient.get<Category>(
      environment.url_product + 'categories/' + id);
  }

  save(category: Category): Observable<Category> {
    if (category.id) {
      return this.httpClient.put<Category>(
        environment.url_product + 'categories', category);
    } else {
      return this.httpClient.post<Category>(
        environment.url_product + 'categories', category);
    }
  }
}
